### PR TITLE
Fixes to L3 Muon OI SeedGenerator (TSGForOI) for improved efficiency

### PR DESCRIPTION
**Pull request by bmahakud**
- The fixes added is supposed to enhance the L3 Outside-in reconstruction efficiency up to 3-4% depending on eta region 
    -  link to slides in HLT meeting : https://indico.cern.ch/event/724303/contributions/2979159/attachments/1638779/2630049/OI_improvementsPurdue_24thApril2018.pdf

- Currently OI starts from L2 muons with
    - Trajectory State on Surface (TSOS) in innermost Muon station (1)
    - TSOS at vertex after updating the parameters with vertex information (2) 
- Fix-I
    - Check compatibility between (1) and (2)
        - 5 parameter Chi2 and DeltaR(OuterTrakerStateInside,OuterTrackerStateOutside) 
        - If the two TSOS are too far apart use both states to start seed generation
   - To use (1)
       - first propagate from muon system to outer tracker surface using SteppingHelixPropagator,
       -  then use 'PropagatorWithMaterialParabolicMf' to propagate to TOB and TEC layers.  (In the previous version 'PropagatorWithMaterial' was used for both propagation steps.)
- Fix-II
   - Seeds are created on the outermost 2 layers:
      - The layer counter was not properly initialized before we proceed to step (2)
      - Has been fixed, initialized to 0 before layers are found for generating seeds using
TSOS at muon station
      - Scale factors have been optimized for different rapidity ranges:
           - Smaller scale factor in Barrel and Larger in the Endcap Use more eta bins 

- Fix-III
  - Optimize eta range where hits on stereo TEC dets are used to build seeds
    -  Before: If the L2 muon is 0.8-1.6
        - Create seeds using the stereo modules
        - Skip the modules without stereo hits
    - Now: 
       - Range is optimized to: (1.1 to 2.0 ) and (2.15 to 2.4)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by JanFSchulte** [comment]
Hi everyone. 
The tests are failing because this PR removes a parameter in the module, but not in the HLT config in the release (which I guess should only be touched by the TSG) What is the correct way to proceed here to make a viable test?

**Comment by perrotta** [comment]
@bmahakud @JanFSchulte : you must add the customization for the removed parameter in HLTrigger/Configuration/python/customizeHLTforCMSSW.py

@Martin-Grunewald @silviodonato

**Comment by Martin-Grunewald** [comment]
...and I assume a backport PR for 101X is needed as well!

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 1]
```diff
@@ -1,425 +1,475 @@
...
+
+  double theSF = 1.0;
+  //	L2 direction vs pT blowup - as was previously done:
+  //	Split into 4 pT ranges: <pT1_, pT1_<pT2_, pT2_<pT3_, <pT4_: 13,30,70
+  //	Split into different eta ranges depending in pT
+
+
+  double abseta = std::abs(track->eta());
+  if (track->pt() <= pT1_) theSF = SF1_;
+  if (track->pt() > pT1_ && track->pt() <= pT2_) {
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: Assuming that all ```pt[i]_``` and ```eta[i]_``` are sorted, here and below all mutually excluded pt/eta ranges should be checked with ```else if```

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 2]
```diff
@@ -1,425 +1,475 @@
...
+				const TrackerTopology* tTopo,
+				const GeometricSearchDet& layer,
+				const TrajectoryStateOnSurface& tsosAtIP,
+				const Propagator& propagatorAlong,
+				const Propagator& propagatorOpposite,
+				const reco::TrackRef l2,
+				edm::ESHandle<Chi2MeasurementEstimatorBase>& estimatorH,
+				edm::Handle<MeasurementTrackerEvent>& measurementTrackerH,
+				unsigned int& numSeedsMade,
+                unsigned int& hitlessSeedsMade,
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: indent as the other lines

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 3]
```diff
@@ -1,425 +1,475 @@
...
+    // ENDCAP-
+    if (l2->eta() < -minEtaForTEC_) {
+      layerCount = 0;
+      for (auto it=tecNegative.rbegin(); it!=tecNegative.rend(); ++it) {
+	    LogTrace("TSGForOI") << "TSGForOI::produce: looping in TEC- layer " << layerCount << endl;
+	    findSeedsOnLayer(tTopo, **it, tsosAtIP,  *(propagatorAlong.get()), *(propagatorOpposite.get()), l2,
+                         estimatorH, measurementTrackerH, numSeedsMade, hitlessSeedsMade, numOfMaxSeeds, layerCount, out);
+	  }
+	  if (useBoth) {
+        layerCount = 0;
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: Indent the whole scope

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 4]
```diff
@@ -1,425 +1,475 @@
...
+    // ENDCAP+
+    if (l2->eta() > minEtaForTEC_) {
+       layerCount = 0;
+      for (auto it=tecPositive.rbegin(); it!=tecPositive.rend(); ++it) {
+	    LogTrace("TSGForOI") << "TSGForOI::produce: looping in TEC+ layer " << layerCount << endl;
+	    findSeedsOnLayer(tTopo, **it, tsosAtIP, *(propagatorAlong.get()), *(propagatorOpposite.get()), l2,
+                         estimatorH, measurementTrackerH, numSeedsMade, hitlessSeedsMade, numOfMaxSeeds, layerCount, out);
+	  }
+	  if (useBoth) {
+        layerCount = 0;
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: indent the whole scope

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 5]
```diff
@@ -1,425 +1,475 @@
...
+
+}
+
+
+//
+// Produce seeds
+//
+void TSGForOI::produce(edm::StreamID sid, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+    
+  using namespace std;
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: Not needed
(You can probably also avoid the one at line 9, simply checking if an explicit ```std::``` is missing somewhere - not at a first glance...)

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 6]
```diff
@@ -1,425 +1,475 @@
...
+                                                                measurementTrackerH->geometricSearchTracker()->negTecLayers();
+  edm::ESHandle<TrackerTopology> tTopo_handle;
+  iSetup.get<TrackerTopologyRcd>().get(tTopo_handle);
+  const TrackerTopology* tTopo = tTopo_handle.product();
+
+  // Get suitable propagators
+  std::unique_ptr<Propagator> propagatorAlong = SetPropagationDirection(*propagatorAlongH,alongMomentum);
+  std::unique_ptr<Propagator> propagatorOpposite = SetPropagationDirection(*propagatorOppositeH,oppositeToMomentum);
+
+  // Stepping Helix Propagator for propogation from uon system to tracker
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: uon -> muon

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 7]
```diff
@@ -1,425 +1,475 @@
...
+      if (useBoth) {
+        layerCount = 0;
+        numSeedsMade = 0;
+        hitlessSeedsMade = 0;
+        for (auto it=tob.rbegin(); it!=tob.rend(); ++it) {	//This goes from outermost to innermost layer
+          LogTrace("TSGForOI") << "TSGForOI::produce: looping in TOB layer " << layerCount << endl;
+          findSeedsOnLayer(tTopo, **it, outerTkStateOutside,  *(propagatorOpposite.get()), *(propagatorOpposite.get()), l2,
+                           estimatorH, measurementTrackerH, numSeedsMade, hitlessSeedsMade, numOfMaxSeeds, layerCount, out);
+        }
+	  }
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: indent

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 8]
```diff
@@ -1,425 +1,475 @@
...
-
-DEFINE_FWK_MODULE(TSGForOI);
+/**
+  \class    TSGForOI
+  \brief    Create L3MuonTrajectorySeeds from L2 Muons updated at vertex in an outside in manner
+  \author   Benjamin Radburn-Smith, Santiago Folgueras
+ */
+
+#include "RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.h"
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: Please remove from hre if you include it already in the header file

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 9]
```diff
@@ -1,425 +1,475 @@
...
+      if (dist1 > tsosDiff1_ || dist2 > tsosDiff2_) useBoth = true;
+    }
+
+    numSeedsMade = 0;
+    hitlessSeedsMade = 0;
+
+    // BARREL
+    if (std::abs(l2->eta()) < maxEtaForTOB_) {
+      layerCount = 0;
+      for (auto it=tob.rbegin(); it!=tob.rend(); ++it) {	// this runs from outermost to innermost TOB layer
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: (Just to check): shouldn't this loop be computed only ``` if (!useBoth```?

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 10]
```diff
@@ -1,425 +1,475 @@
...
+    }
+    // Reset Number of seeds if in overlap region
+    if (std::abs(l2->eta())>minEtaForTEC_ && std::abs(l2->eta())<maxEtaForTOB_){
+       numSeedsMade = 0;
+    }
+
+    // ENDCAP+
+    if (l2->eta() > minEtaForTEC_) {
+       layerCount = 0;
+      for (auto it=tecPositive.rbegin(); it!=tecPositive.rend(); ++it) {
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: (Just to check): shouldn't this loop be computed only ``` if (!useBoth```?

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 11]
```diff
@@ -1,425 +1,475 @@
...
+	    LogTrace("TSGForOI") << "TSGForOI::produce: looping in TEC+ layer " << layerCount << endl;
+	    findSeedsOnLayer(tTopo, **it, tsosAtIP, *(propagatorAlong.get()), *(propagatorOpposite.get()), l2,
+                         estimatorH, measurementTrackerH, numSeedsMade, hitlessSeedsMade, numOfMaxSeeds, layerCount, out);
+	  }
+	  if (useBoth) {
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: indent

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 12]
```diff
@@ -1,425 +1,475 @@
...
+          findSeedsOnLayer(tTopo, **it, outerTkStateOutside,  *(propagatorOpposite.get()), *(propagatorOpposite.get()), l2,
+                           estimatorH, measurementTrackerH, numSeedsMade, hitlessSeedsMade, numOfMaxSeeds, layerCount, out);
+		}
+      }
+    }
+
+    // ENDCAP-
+    if (l2->eta() < -minEtaForTEC_) {
+      layerCount = 0;
+      for (auto it=tecNegative.rbegin(); it!=tecNegative.rend(); ++it) {
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: (Just to check): shouldn't this loop be computed only ``` if (!useBoth```?

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 13]
```diff
@@ -1,425 +1,475 @@
...
+	    LogTrace("TSGForOI") << "TSGForOI::produce: looping in TEC- layer " << layerCount << endl;
+	    findSeedsOnLayer(tTopo, **it, tsosAtIP,  *(propagatorAlong.get()), *(propagatorOpposite.get()), l2,
+                         estimatorH, measurementTrackerH, numSeedsMade, hitlessSeedsMade, numOfMaxSeeds, layerCount, out);
+	  }
+	  if (useBoth) {
```
Comment (perrotta) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: indent

**Review by perrotta — COMMENTED** [review]
While adding the customization for HLT, please profit and fix what pointed out in the following comments. Probably a few more will follow once the code will become better readable.

**Comment by bmahakud** [comment]
Hi @perrotta  and all ,

Thanks. We will fix it soon as possible.

**Comment by perrotta** [comment]
@bmahakud : any news about the fixes?

**Comment by bmahakud** [comment]
Dear @perrotta ,

It has been observed that the new OI seed generator fixes has increased the timing slightly even though it increases the efficiency. We have been asked to add more fixes so that it reduces the timing back to the default code. We are working on it. Hope to have the fix in next 2-3 days. We will have to update the pull request again after we have the fix for timing.
Bibhu

**Comment by perrotta** [comment]
@bmahakud  (just checking): any update?

**Comment by perrotta** [comment]
-1
This needs to be fixed anyhow, and apparently there are no progresses since a while.
Let move it out from our list, then: we will continue the review if and when there will be any update

**Comment by bmahakud** [comment]
Hi @perrotta ,

I will close the pull request for the moment and reopen after incorporating the changes. I still need couple of days more to reduce make the code ready. Is this okay?

**Comment by perrotta** [comment]
Thank you @bmahakud: we can resume the review as soon as you will update it.
Let us know when you are done with it.

bmahakud <notifications@github.com> ha scritto:

**Pull request closed without merging**